### PR TITLE
Public developer guide

### PR DESCRIPTION
**Pull request by stephanielee9**
Creating a client-side dev guide

### CODE REVIEW

File: docs/source/resources/developer_guide.rst [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+Installation
+------------
+Please visit our `README <https://github.com/Infleqtion/client-superstaq/blob/main/README.md#installation-for-development>`_ for developer installation instructions.
+
+Modifying Superstaq
+-------------------
+
+Push Edits and Make a Pull Request for Review
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#. Update the main branch (e.g., ``git pull``) and update relevant dependencies (``python3 -m pip install -e .``, ``python3 -m pip install <general-superstaq, qiskit-superstaq, cirq-superstaq. --upgrade``).
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:24]: this install line isn't quite right - should be something like,
```suggestion
#. Update the main branch (e.g., ``git pull``)
#. Install/update relevant dependencies. To work on a single package, navigate to relevant subdirectory (e.g. `cd cirq-superstaq`) and run ``python3 -m pip install -e .``. Alternatively, install all ``client-superstaq`` packages by navigating to the repository root and running ``python3 -m pip install -e ./checks-superstaq -e ./general-superstaq -e ./qiskit-superstaq -e ./cirq-superstaq -e ./supermarq-benchmarks`` (in that order).
```
(though probably cleaned up a bit)

Comment (stephanielee9) [reply docs/source/resources/developer_guide.rst]: Add the ^ language with some formatting - let me know how it looks and if you want me to update anything!

File: docs/source/resources/developer_guide.rst [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+|   from qiskit import pulse                       |   import qiskit                                  |
+|   with pulse.build(backend) as program:          |   with qiskit.pulse.build(backend) as program:   |
+|       gaussian_pulse = library.gaussian(...)     |       gaussian_pulse = library.gaussian(...)     |
+|       pulse.play(gaussian_pulse)                 |       pulse.play(gaussian_pulse)                 |
++--------------------------------------------------+--------------------------------------------------+
+
+
+Exceptions can be made if the explicit name is extremely long (e.g. ``from qiskit.providers import JobStatus as qjs``) or if the name is used extremely frequently and has low risk of name clash (e.g. ``from typing import Iterable, Union``).
+
+We allow ``web`` to import ``superstaq``, but not the reverse. Similarly, we allow ``dev_tools`` to import ``web``, but not the reverse.
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:201]:
```suggestion
```

File: docs/source/resources/developer_guide.rst [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+Productivity Tips
+=================
+* For command-line python testing, use the ``ipython`` shell instead of normal python. ``ipython`` is like Jupyter but within your terminal.
+* For either ``ipython`` or Jupyter, consider starting with
+
+  >>> %load_ext autoreload
+  >>> %autoreload 2
+
+  This will enable autoreload.
+* Superstaq/dev_tools/pygrep.sh contains a nice script for grepping python files. Recommendation: alias ``p`` to this script (by adding ``alias p='/path/to/Superstaq/dev_tools/pygrep.sh'`` to ~/.zprofile or wherever aliases should be stored). Then you can type something like ``p AQTDevice``
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:213]:
```suggestion
```
(`pygrep.sh` isn't in client-superstaq)

File: docs/source/resources/developer_guide.rst [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+.. code-block:: bash
+
+   alias ss='cd /path/to/Superstaq'
+   alias py='ipython'
+   alias pyr="ipython --InteractiveShellApp.extensions 'autoreload' --InteractiveShellApp.exec_lines '%autoreload 2'"
+
+(That last one, ``pyr``, runs an ipython shell with autoreloading already set up.)
+
+* View your AWS credentials with ``cat ~/.aws/credentials``, and store them in environment variables with, e.g., ``export AWS_ACCESS_KEY_ID=...``.
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:224]:
```suggestion
```

File: docs/source/resources/developer_guide.rst [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+  >>> %load_ext autoreload
+  >>> %autoreload 2
+
+  This will enable autoreload.
+* Superstaq/dev_tools/pygrep.sh contains a nice script for grepping python files. Recommendation: alias ``p`` to this script (by adding ``alias p='/path/to/Superstaq/dev_tools/pygrep.sh'`` to ~/.zprofile or wherever aliases should be stored). Then you can type something like ``p AQTDevice``
+* Also recommended: 
+
+.. code-block:: bash
+
+   alias ss='cd /path/to/Superstaq'
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:218]:
```suggestion
```

File: docs/source/resources/developer_guide.rst [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+   alias py='ipython'
+   alias pyr="ipython --InteractiveShellApp.extensions 'autoreload' --InteractiveShellApp.exec_lines '%autoreload 2'"
+
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:221]:
```suggestion
alias pygrep="grep -r --color=auto --include='*.py'"

```
(this does the same thing as `pygrep.sh`)

File: docs/source/resources/developer_guide.rst
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:1]: `check/` should be replaced with `checks/` everywhere it appears in this file

File: docs/source/resources/developer_guide.rst [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+===========
+
+Nomenclature and Abbreviations
+------------------------------
+
+Operations can be referred to as ``op``.
+
+``target`` should always be used to refer to device names (type ``str``) in the Superstaq "<vendor>_<device>_<type>" format, e.g. "ibmq_qasm_simulator".  The same will be used accross all clients. ``backend_name`` should be used for vendor-specific names of particular hardware backends, and ``backend`` for actual instances thereof.
+
+Circuits
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:101]: this is cirq-specific - maybe we should change the header to
```suggestion
Cirq
```
?

File: docs/source/resources/developer_guide.rst (hunk 1 above)
Comment (richrines1) [reply docs/source/resources/developer_guide.rst]: perfect, thanks!

File: docs/source/resources/developer_guide.rst
Comment (stephanielee9) [reply docs/source/resources/developer_guide.rst:1]: updated!

**Comment by stephanielee9** [comment]
@richrines1 I think all your comments are implemented now!

File: docs/source/resources/developer_guide.rst [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+  This will enable autoreload.
+* Also recommended: 
+
+.. code-block:: bash
+
+   alias py='ipython'
+   alias pyr="ipython --InteractiveShellApp.extensions 'autoreload' --InteractiveShellApp.exec_lines '%autoreload 2'"
+alias pygrep="grep -r --color=auto --include='*.py'"
+
+(That last one, ``pyr``, runs an ipython shell with autoreloading already set up.)
```
Comment (richrines1) [inline docs/source/resources/developer_guide.rst:219]: nit: `pyr` isn't the "last one" anymore (maybe just reorder them?)

**Review by richrines1 — APPROVED** [review]
lgtm % nit!

**Pull request merged**